`contract3` for symmetric tensors

Currently, the `contract3` function which computes the product of 3 tensors is only defined for `Tensor`, not for `SymmetricTensor`. Adding this function would make general and symmetric tensors more consistent. It might also be a performance gain over computing the product using the overloaded `*` operator, which will create a temporary value.

@tamiko I'm busy working on adding this functionality using the suggestion you made. I've hit a snag that within `TensorAccessors::contract3` one needs to retrieve a tensor row with `operator[]`. However, this does not seem to work for `SymmetricTensor`s:
```
485: .../deal.II/base/tensor_accessors.h:745:92: error: no viable overloaded operator[] for type 'const dealii::internal::SymmetricTensorAccessors::Accessor<2, 3, true, 1, double>'
485:           result += Contract3<0, rank_2 - 1, dim>::template contract3<T1>(left, middle[i], right[i]);
```
This is understandable as the data structures for the `Tensor` and `Symmetric` tensor classes are different. I could write a function like `extract_subtensor(right,i)` that calls `operator[]` for a `Tensor` object but builds the necessary data for the `SymmetricTensor` object, but this seems a bit wasteful since it'll mean creating a temporary `Tensor<1,dim>` when `right` is a `SymmetricTensor<2,dim>`.

Do you perhaps have any other suggestions?

--- a/doc/news/changes/minor/20170110Jean-PaulPelteretMatthiasMaier
+++ b/doc/news/changes/minor/20170110Jean-PaulPelteretMatthiasMaier
@@ -1,0 +1,4 @@
+Improved: The contract3() function has been extended accommodate both Tensor
+and SymmetricTensors arguments.
+<br>
+(Jean-Paul Pelteret, Matthias Maier, 2017/01/10)

--- a/include/deal.II/base/symmetric_tensor.h
+++ b/include/deal.II/base/symmetric_tensor.h
@@ -323,6 +323,11 @@ namespace internal
        */
       Accessor<rank,dim,constness,P-1,Number> operator [] (const unsigned int i);
 
+      /**
+       * Index operator.
+       */
+      Accessor<rank,dim,constness,P-1,Number> operator [] (const unsigned int i) const;
+
     private:
       /**
        * Store the data given to the constructor.
@@ -405,6 +410,11 @@ namespace internal
        * Index operator.
        */
       reference operator [] (const unsigned int);
+
+      /**
+       * Index operator.
+       */
+      reference operator [] (const unsigned int) const;
 
     private:
       /**
@@ -867,6 +877,16 @@ namespace internal
 
 
 
+    template <int rank, int dim, bool constness, int P, typename Number>
+    Accessor<rank,dim,constness,P-1,Number>
+    Accessor<rank,dim,constness,P,Number>::operator[] (const unsigned int i) const
+    {
+      return Accessor<rank,dim,constness,P-1,Number> (tensor,
+                                                      merge (previous_indices, i, rank-P));
+    }
+
+
+
     template <int rank, int dim, bool constness, typename Number>
     Accessor<rank,dim,constness,1,Number>::
     Accessor (tensor_type              &tensor,
@@ -891,6 +911,14 @@ namespace internal
     template <int rank, int dim, bool constness, typename Number>
     typename Accessor<rank,dim,constness,1,Number>::reference
     Accessor<rank,dim,constness,1,Number>::operator[] (const unsigned int i)
+    {
+      return tensor(merge (previous_indices, i, rank-1));
+    }
+
+
+    template <int rank, int dim, bool constness, typename Number>
+    typename Accessor<rank,dim,constness,1,Number>::reference
+    Accessor<rank,dim,constness,1,Number>::operator[] (const unsigned int i) const
     {
       return tensor(merge (previous_indices, i, rank-1));
     }

--- a/include/deal.II/base/tensor.h
+++ b/include/deal.II/base/tensor.h
@@ -1445,10 +1445,10 @@ operator- (const Tensor<rank,dim,Number> &p, const Tensor<rank,dim,OtherNumber> 
  * index of a tensor @p src1 of rank @p rank_1 with the first index of a
  * tensor @p src2 of rank @p rank_2:
  * @f[
- *   \text{result}_{i_1,..,i_{r1},j_1,..,j_{r2}}
+ *   \text{result}_{i_1,\ldots,i_{r1},j_1,\ldots,j_{r2}}
  *   = \sum_{k}
- *     \text{left}_{i_1,..,i_{r1}, k}
- *     \text{right}_{k, j_1,..,j_{r2}}
+ *     \text{left}_{i_1,\ldots,i_{r1}, k}
+ *     \text{right}_{k, j_1,\ldots,j_{r2}}
  * @f]
  *
  * @note For the Tensor class, the multiplication operator only performs a
@@ -1485,10 +1485,10 @@ operator * (const Tensor<rank_1, dim, Number> &src1,
  * contraction of index @p index_1 of a tensor @p src1 of rank @p rank_1 with
  * the index @p index_2 of a tensor @p src2 of rank @p rank_2:
  * @f[
- *   \text{result}_{i_1,..,i_{r1},j_1,..,j_{r2}}
+ *   \text{result}_{i_1,\ldots,i_{r1},j_1,\ldots,j_{r2}}
  *   = \sum_{k}
- *     \text{left}_{i_1,..,k,..,i_{r1}}
- *     \text{right}_{j_1,..,k,..,j_{r2}}
+ *     \text{left}_{i_1,\ldots,k,\ldots,i_{r1}}
+ *     \text{right}_{j_1,\ldots,k,\ldots,j_{r2}}
  * @f]
  *
  * If for example the first index (<code>index_1==0</code>) of a tensor
@@ -1546,10 +1546,10 @@ contract (const Tensor<rank_1, dim, Number> &src1,
  * index_3 with index @p index_4 of a tensor @p src1 of rank @p rank_1 and a
  * tensor @p src2 of rank @p rank_2:
  * @f[
- *   \text{result}_{i_1,..,i_{r1},j_1,..,j_{r2}}
+ *   \text{result}_{i_1,\ldots,i_{r1},j_1,\ldots,j_{r2}}
  *   = \sum_{k, l}
- *     \text{left}_{i_1,..,k,..,l,..,i_{r1}}
- *     \text{right}_{j_1,..,k,..,l..,j_{r2}}
+ *     \text{left}_{i_1,\ldots,k,\ldots,l,\ldots,i_{r1}}
+ *     \text{right}_{j_1,\ldots,k,\ldots,l\ldots,j_{r2}}
  * @f]
  *
  * If for example the first index (<code>index_1==0</code>) shall be
@@ -1626,9 +1626,9 @@ double_contract (const Tensor<rank_1, dim, Number> &src1,
  * of equal rank: Return a scalar number that is the result of a full
  * contraction of a tensor @p left and @p right:
  * @f[
- *   \sum_{i_1,..,i_r}
- *   \text{left}_{i_1,..,i_r}
- *   \text{right}_{i_1,..,i_r}
+ *   \sum_{i_1,\ldots,i_r}
+ *   \text{left}_{i_1,\ldots,i_r}
+ *   \text{right}_{i_1,\ldots,i_r}
  * @f]
  *
  * @relates Tensor
@@ -1652,10 +1652,10 @@ scalar_product (const Tensor<rank, dim, Number> &left,
  * tensor @p middle of rank $(\text{rank}_1+\text{rank}_2)$ and a tensor @p
  * right of rank @p rank_2:
  * @f[
- *   \sum_{i_1,..,i_{r1},j_1,..,j_{r2}}
- *   \text{left}_{i_1,..,i_{r1}}
- *   \text{middle}_{i_1,..,i_{r1},j_1,..,j_{r2}}
- *   \text{right}_{j_1,..,j_{r2}}
+ *   \sum_{i_1,\ldots,i_{r1},j_1,\ldots,j_{r2}}
+ *   \text{left}_{i_1,\ldots,i_{r1}}
+ *   \text{middle}_{i_1,\ldots,i_{r1},j_1,\ldots,j_{r2}}
+ *   \text{right}_{j_1,\ldots,j_{r2}}
  * @f]
  *
  * @note Each of the three input tensors can be either a Tensor or
@@ -1685,8 +1685,8 @@ contract3 (const TensorT1<rank_1, dim, T1>          &left,
  * The outer product of two tensors of @p rank_1 and @p rank_2: Returns a
  * tensor of rank $(\text{rank}_1 + \text{rank}_2)$:
  * @f[
- *   \text{result}_{i_1,..,i_{r1},j_1,..,j_{r2}}
- *   = \text{left}_{i_1,..,i_{r1}}\,\text{right}_{j_1,..,j_{r2}.}
+ *   \text{result}_{i_1,\ldots,i_{r1},j_1,\ldots,j_{r2}}
+ *   = \text{left}_{i_1,\ldots,i_{r1}}\,\text{right}_{j_1,\ldots,j_{r2}.}
  * @f]
  *
  * @relates Tensor

--- a/include/deal.II/base/tensor.h
+++ b/include/deal.II/base/tensor.h
@@ -1658,15 +1658,21 @@ scalar_product (const Tensor<rank, dim, Number> &left,
  *   \text{right}_{j_1,..,j_{r2}}
  * @f]
  *
+ * @note Each of the three input tensors can be either a Tensor or
+ * SymmetricTensor.
+ *
  * @relates Tensor
- * @author Matthias Maier, 2015
+ * @author Matthias Maier, 2015, Jean-Paul Pelteret 2017
  */
-template <int rank_1, int rank_2, int dim,
+template <template<int, int, typename> class TensorT1,
+          template<int, int, typename> class TensorT2,
+          template<int, int, typename> class TensorT3,
+          int rank_1, int rank_2, int dim,
           typename T1, typename T2, typename T3>
 typename ProductType<T1, typename ProductType<T2, T3>::type>::type
-contract3 (const Tensor<rank_1, dim, T1> &left,
-           const Tensor<rank_1 + rank_2, dim, T2> &middle,
-           const Tensor<rank_2, dim, T3> &right)
+contract3 (const TensorT1<rank_1, dim, T1>          &left,
+           const TensorT2<rank_1 + rank_2, dim, T2> &middle,
+           const TensorT3<rank_2, dim, T3>          &right)
 {
   typedef typename ProductType<T1, typename ProductType<T2, T3>::type>::type
   return_type;

--- a/tests/base/symmetric_tensor_37.cc
+++ b/tests/base/symmetric_tensor_37.cc
@@ -1,0 +1,87 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2017 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+
+// Check the contract3 function involving SymmetricTensors
+
+#include "../tests.h"
+#include <deal.II/base/symmetric_tensor.h>
+#include <deal.II/base/tensor.h>
+#include <deal.II/base/logstream.h>
+
+#include <fstream>
+#include <iomanip>
+
+// Although not strictly required, in this test it is expected that T1 or T3
+// (or both) are a symmetric tensor
+template<int rank1, int rank2, int rank3, int dim, typename number,
+         template<int,int,typename> class T1,
+         template<int,int,typename> class T3>
+void test_symm_tensor_contract_3(const T1<rank1,dim,number>     &l,
+                                 const Tensor<rank2,dim,number> &m,
+                                 const T3<rank3,dim,number>     &r)
+{
+  const double res1 = contract3(l,m,r);
+  const double res2 = contract3(static_cast<Tensor<rank1,dim> >(l),
+                                m,
+                                static_cast<Tensor<rank3,dim> >(r));
+  Assert(std::abs(res1 - res2) < 1e-12,
+         ExcMessage("Result from symmetric tensor contract3 is incorrect."));
+}
+
+int main ()
+{
+  std::ofstream logfile("output");
+  deallog << std::setprecision(5);
+  deallog.attach(logfile);
+  deallog.threshold_double(1.e-10);
+
+  const int dim = 3;
+  Tensor<1,dim> v1;
+  Tensor<2,dim> t1;
+  SymmetricTensor<2,dim> s1,s2;
+  Tensor<3,dim> T1;
+  Tensor<4,dim> H1;
+
+  for (unsigned int i=0; i<dim; ++i)
+    {
+      v1[i] = 1+i;
+      for (unsigned int j=0; j<dim; ++j)
+        {
+          t1[i][j] = 1+j+i*dim;
+          s1[i][j] = 1+j+i*dim;
+          s2[i][j] = 2+(j+i)*dim;
+          for (unsigned int k=0; k<dim; ++k)
+            {
+              T1[i][j][k] = 1+k+j*dim+i*dim*dim;
+              for (unsigned int l=0; l<dim; ++l)
+                H1[i][j][k][l] = 1+l+k*dim+j*dim*dim+i*dim*dim*dim;
+            }
+        }
+    }
+
+  // Vector + SymmetricTensor
+  test_symm_tensor_contract_3(v1,T1,s1);
+  test_symm_tensor_contract_3(s1,T1,v1);
+
+  // Tensor + SymmetricTensor
+  test_symm_tensor_contract_3(t1,H1,s1);
+  test_symm_tensor_contract_3(s1,H1,t1);
+
+  // SymmetricTensor + SymmetricTensor
+  test_symm_tensor_contract_3(s1,H1,s2);
+
+  deallog << "All OK" << std::endl;
+}

--- a/tests/base/symmetric_tensor_37.output
+++ b/tests/base/symmetric_tensor_37.output
@@ -1,0 +1,2 @@
+
+DEAL::All OK

